src/kernel/MomentumOpenAdvDiffElemKernel.C

```diff
@@ -0,0 +1,370 @@
+    const DoubleType densL   = *stk::mesh::field_data(densityNp1, nodeL);
+    const DoubleType densR   = *stk::mesh::field_data(densityNp1, nodeR);
+    const DoubleType viscL   = *stk::mesh::field_data(*viscosity_, nodeL);
+    const DoubleType viscR   = *stk::mesh::field_data(*viscosity_, nodeR);
+    const DoubleType *uNp1R  =  stk::mesh::field_data(velocityNp1, nodeR);
+    const DoubleType *vrtmL  =  stk::mesh::field_data(*velocityRTM_, nodeL);
+    const DoubleType *vrtmR  =  stk::mesh::field_data(*velocityRTM_, nodeR);
+    
+    const DoubleType *coordL =  stk::mesh::field_data(*coordinates_, nodeL);
+    const DoubleType *coordR =  stk::mesh::field_data(*coordinates_, nodeR);
+    
+    DoubleType udotx = 0.0;
+    for ( int i = 0; i < BcAlgTraits::nDim_; ++i ) {
+      const DoubleType dxi = coordR[i]  - coordL[i];
+      udotx += 0.5*dxi*(vrtmL[i] + vrtmR[i]);
+      w_nx[i] = vf_exposedAreaVec(ip,i)/amag;
+      // extrapolation
+      DoubleType duR = 0.0;
+      for ( int j = 0; j < BcAlgTraits::nDim_; ++j ) {
+        DoubleType dxj = w_coordBip[j] - coordR[j];
+        duR += dxj*vf_Gjui(localFaceNode,i,j)*hoUpwind;
+      }
+      w_uBipExtrap[i] = uNp1R[i] + duR;
+    }
+    
+    const DoubleType diffIp = 0.5*(viscL/densL + viscR/densR);
+    const DoubleType pecFuncArg = stk::math::abs(udotx)/(diffIp+small_);
+    // FIXME: modify pecletFunction to be double type
+    DoubleType pecfac = 0.0;
+    for(int simdIndex=0; simdIndex<stk::simd::ndoubles; ++simdIndex) {
+      stk::simd::set_data(pecfac, simdIndex, pecletFunction_->execute(stk::simd::get_data(pecFuncArg, simdIndex)));
+    }
+    const DoubleType om_pecfac = 1.0-pecfac;
+   
+    //================================
+    // advection first
+    //================================
+    const DoubleType tmdot = vf_openMassFlowRate(ip);
+    
+    // advection; leaving the domain
+    // fixme... oh my, do I need to wrap this in a stk::math::if_then_else(tmdot > 0, doThis, doThat);
+    if ( tmdot > 0.0 ) {
+      
+      for ( int i = 0; i < BcAlgTraits::nDim_; ++i ) {
+        
+        const int indexR = nearestNode*BcAlgTraits::nDim_ + i;
+        
+        // central
+        const DoubleType uiIp = w_uBip[i];
+        
+        // upwind
+        const DoubleType uiUpwind = alphaUpw*w_uBipExtrap[i] + om_alphaUpw*uiIp;
+        
+        // total advection; pressure contribution in time expression
+        const DoubleType aflux = tmdot*(pecfac*uiUpwind+om_pecfac*uiIp);
+        
+        rhs(indexR) -= aflux;
+        
+        // upwind lhs
+        lhs(indexR, nearestNode*BcAlgTraits::nDim_+i) += tmdot*pecfac*alphaUpw;
+        
+        // central part
+        const DoubleType fac = tmdot*(pecfac*om_alphaUpw+om_pecfac);
+        for ( int ic = 0; ic < NcAlgTraits::nodesPerFace_; ++ic ) {
+          const DoubleType r = vf_shape_function(ip,ic);
+          // fixme: where to extract face_node_ordinals
+          const int nn = face_node_ordinals[ic];
```

Comment:
If we need face-node-ordinals here, we may need to consider passing in some struct similar to the SharedMemData struct that is held in run_face_elem_algorithm but which is currently not passed down to kernel->execute. Currently that struct has elem-face-ordinals. It has a place-holder for connectedNodes but that isn't yet populated. We'll have to discuss just how much stuff we want(need) to populate and pass down.

Reply:
I think that @rcknaus is dealing with this for his master element exposed face grad_op... Thoughts Robert?